samza-core/src/main/java/org/apache/samza/operators/StreamGraphImpl.java

```diff
@@ -156,4 +161,46 @@ public ContextManager getContextManager() {
   /* package private */ int getNextOpId() {
     return this.opId++;
   }
+
+  /**
+   * Get all {@link OperatorSpec}s available in this {@link StreamGraphImpl}
+   *
+   * @return  a set of all available {@link OperatorSpec}s
+   */
+  public Set<OperatorSpec> getAllOperatorSpecs() {
+    Collection<InputStreamInternal> inputStreams = inStreams.values();
+    Set<OperatorSpec> operatorSpecs = new HashSet<>();
+
+    for (InputStreamInternal stream : inputStreams) {
+      doGetOperatorSpecs((MessageStreamImpl) stream, operatorSpecs);
+    }
+    return operatorSpecs;
+  }
+
+  private void doGetOperatorSpecs(MessageStreamImpl stream, Set<OperatorSpec> specs) {
+    Collection<OperatorSpec> registeredOperatorSpecs = stream.getRegisteredOperatorSpecs();
+    for (OperatorSpec spec : registeredOperatorSpecs) {
+      specs.add(spec);
+      MessageStreamImpl nextStream = spec.getNextStream();
+      if (nextStream != null) {
+        //Recursively traverse and obtain all reachable operators
+        doGetOperatorSpecs(nextStream, specs);
+      }
+    }
+  }
+
+  /**
+   * Returns <tt>true</tt> iff this {@link StreamGraphImpl} contains a join or a window operator
+   *
+   * @return  <tt>true</tt> iff this {@link StreamGraphImpl} contains a join or a window operator
+   */
+  public boolean hasWindowOrJoins() {
```

Comment:
+1 for naming.

----

Not necessary in this PR, but would it be cleaner to consolidate all the logic in one place, either the planner or here? For example, StreamGraph could have a single public method `getTimerInterval` (default = -1) that does the hasTimedOperators + traversal + gcd calculation, and the logic in Planner could be:

```
long timerInterval = graph.getTimerInterval();
if (windowMs == null || windowMs <= 0) {
  LOG.info("...");
  windowMs = timerInterval;
} else if (windowMs > timerInterval) {
  LOG.warn("...");
  windowMs = timerInterval; // maybe 
}
```